bayesian_inversion.m

```diff
@@ -0,0 +1,211 @@
+    %  Let's assume X0 and Xpr are the same.
+    phi0 = 0.01; % melt fraction
+    g0 = log10(1000.); % 1 mm (1000 microns)
+    % g0 = 4000.; % 5 mm
+
+    [X0,iT,iphi,ig] = make_X0_xfit_premelt(vs,nz,nlat,nlon,nmod,npts,vs_vbr,phi0,g0,phi,g,T);
+    Xpr = X0;
+
+    %  (2) Frechet kernel (F) of the starting model
+    F = make_F(vs_vbr,lQ_vbr,nlat,nlon,nz,ndata,nmod,T,phi,g,iT,iphi,ig,nT,nphi,ng);
+
+    %  (3) Make Covariance Matrices
+    %      Data covariance matrix (Vd):
+    Vd = make_Vd(ndata,nlat,nlon,nz,lQ_err,vs_err);
+    %      Model covariance matrix (Vm):
+    std_T = ones(npts)*50.;
+    std_phi = ones(npts)*0.0025;
+    std_g = ones(npts)*0.2; % log units
+    % std_g = ones(npts)*2000; % linear units (microns)
+    lscale = 200.; % km
+    Vm = make_Vm(std_T,std_phi,std_g,lscale,lats,lons,zs,npts,nmod);
+
+    %  (4) Residual (y - f(X))
+    yres = calc_yres(ndata,nlat,nlon,nz,iT,iphi,ig,vs_vbr,lQ_vbr,vs,lQ);
+
+    %% BEGIN INVERSION
+
+    %  Only Xk, F, and yres are updated in this scheme...
+
+    irun = 1;
+    % invert some matrices:
+    Vdi = Vd\eye(ndata);
+    Vmi = Vm\eye(nmod);
+    Xk = X0;
+
+    iter=1;
+
+    iphi_idx = (2:3:nmod);
+    iT_idx = (1:3:nmod);
+    ig_idx = (3:3:nmod);
+
+    while irun == 1
+        
+        % form inversion
+
+        mat1 = F'*Vdi*F;
+        mat2 = Vmi;
+        matA = mat1+mat2;
+        matA = matA\eye(nmod);
+        
+        mat1 = F'*Vdi*yres;
+        mat2 = Vmi*(Xk-Xpr);
+        matB = mat1-mat2;
+        
+        Xk1 = Xk + matA*matB;
+        
+        iphi0s = find(Xk1(iphi_idx)<0); % can't have negative melt
+        Xk1(iphi_idx(iphi0s))=0.;
+        
+    %     ig0s = find(Xk1(ig_idx)<0.1e3); % can't have small grain size
+    %     Xk1(ig_idx(ig0s))=0.1e3;
+    %    
+        % calculate misfit    
+        [iTs,iphis,igs] = find_idx(Xk1,npts,nlat,nlon,nz,phi,g,T);
+        yres = calc_yres(ndata,nlat,nlon,nz,iTs,iphis,igs,vs_vbr,lQ_vbr,vs,lQ);
+        chi2 = yres'*Vdi*yres/ndata;
+        display(['Chi-Squared ',num2str(chi2),' for iteration ',num2str(iter)])
+
+        max_iter = 1;
+        chi_limit = 1;
+        if ((chi2 > chi_limit)&&(iter<max_iter))
+            irun=1;
+            
+            % update F
+            F = make_F(vs_vbr,lQ_vbr,nlat,nlon,nz,ndata,nmod,...
+                T,phi,g,iTs,iphis,igs,nT,nphi,ng);
+
+            % update model Xk
+            Xk = Xk1;
+            iter=iter+1;
+        else 
+            irun=0;
+            if (iter>=max_iter)
+                disp(['Did not converge after ' num2str(iter) ' iterations']);
+            else
+                disp(['Converged to chi-squared < ' num2str(chi_limit) ' after ' num2str(iter) ' iterations']);
+            end
+        end
+    end
+
+    %% EXTRACT OUTPUT
+
+    % Calculate the posterior distribution.
+    mat1 = F'*Vdi*F;
+    mat2 = Vmi;
+    matA = mat1+mat2;
+    Vpo = matA\eye(nmod);
+    Vpo_var = diag(Vpo);
+
+    % Extract covariances for chosen sites
```

Comment:
should think about which of the code below this point is necessary for the repository.

Reply:
added a TO DO comment